test-data/unit/check-narrowing.test

```diff
@@ -1269,3 +1269,101 @@ def g() -> None:
         def foo(): ...
     foo()
 [builtins fixtures/dict.pyi]
+
+
+[case testNarrowingIsFinalType]
+from typing import Type, Union
+from typing_extensions import final
+
+@final
+class Mark: ...
+
+@final
+class Other: ...
+
+x: Union[Type[Mark], Type[Other], Type[None], int]
+
+if x is Mark:
```

Comment:
What about `M = NewType("M", Mark")`? This acts as a subtype of `Mark` (and is allowed even though `Mark` is marked `@final`), but `x is Mark` will evaluate to false if `x` is `M`. Maybe that's an edge case that can be ignored, but it does represent a small hole.

Reply:
I guess we can ignore this edge case. I'd be surprised if users are passing the `NewType` definition somewhere as a value, and even more in this particular situation combined with a `@final`. The benefit from this feature outweighs  supporting the edge case.

And even then, isn't the issue only in the opposite case, when we have `if x is not Marc:`? A situation where at runtime we do enter the block, but we narrowed incorrectly.

Reply:
Why can you create a `NewType` of a `final` type? That seems inconsistent - `NewType` creates a subtype, but `final` means that the type isn't allowed to have subclasses. I don't see this mentioned anywhere in the docs/PEPs. I guess it could mean that you have a subset of values, but they all will be the same type in reality. It's still a little contradictory though.